splib/animation/animate.py

```diff
@@ -0,0 +1,231 @@
+    """
+    def __init__(self, *args, **kwargs):
+        Sofa.Core.Controller.__init__(self, *args, **kwargs)
+        self.listening = True
+        self.name = "AnimationManager"
+        self.totalTime = 0
+        self.animations = []
+        self.currentTime = 0
+        self.realTimeStep = None
+
+    def addAnimation(self, animation):
+        self.animations.append(animation)
+
+    def removeAnimation(self, animation):
+        self.animations.remove(animation)
+
+    def bwdInitGraph(self, root):
+        self.onAnimateBeginEvent(0.0)
+
+    def onAnimateBeginEvent(self, event):
+        if self.realTimeStep is not None:
+            self.currentTime = time()
+        self.totalTime = self.totalTime + float(event.get('dt', None))
+        nextanimations = []
+        for animation in self.animations:
+            animation.update(self.totalTime)
+            if animation.factor < 1.0 and animation.direction > 0.0:
+                nextanimations.append(animation)
+            elif animation.factor > 0.0 and animation.direction < 0.0:
+                nextanimations.append(animation)
+            elif animation.mode == "pingpong":
+                animation.direction = -animation.direction
+                animation.startTime = None
+                nextanimations.append(animation)
+            elif animation.mode == "loop":
+                animation.direction = animation.direction
+                animation.startTime = None
+                nextanimations.append(animation)
+            elif animation.onDone is not None:
+                animation.doOnDone(self.totalTime)
+        self.animations = nextanimations
+        return 0
+
+    def onAnimateEndEvent(self, event):
+        timeLeft = self.realTimeStep - (time() - self.currentTime)
+        if self.realTimeStep is not None:
+            sleep(timeLeft)
+
+
+manager = None
+
+def addAnimation(onUpdate, params, duration, mode="once", onDone=None):
+    """Construct and add an animation
+
+    Build a new animation from a callback function that computes the animation value,
+    a set of parameters, the animation duration and the type of animation repetition pattern.
+
+    Animation can be added from any code location (createScene, PythonScriptController)
+
+    :param float duration: duration of the animation in seconds.
+    :param str mode: once, loop, pingpong
+
+    Example:
+        .. sourcecode:: python
+
+            def myAddAnimation(target, factor):
+                print("I should do something on: "+target.name)
+
+
+            def createScene(rootNode)
+                AnimationManager(rootNode)
+                addAnimation(myAddAnimation, {"target" : rootNode }, 10)
+    """
+    if manager == None:
+        raise Exception("Missing manager in this scene")
+    manager.addAnimation(Animation(duration=duration, mode=mode, onUpdate=onUpdate, params=params, onDone=onDone))
+
+def realTimeAnimation(node, step, totalDuration):
```

Comment:
Well I'm not in favor of having this function here. It behave as a kind of animation/simulation loop which is not the purpose of the splib.animation library.